### ISSUE TITLE
Results between vmapped function and jax.lax.map of vmapped function differ significantly depending on GPU and random chance

### ISSUE THREAD

**Issue by SimonKoop** [jax-ml/jax#31919]
My apologies for the very long bug report. The TL;DR is: when I apply the same neural network to the same inputs in two ways
1. on all inputs at once using `jax.vmap`
2. on batches of inputs using `jax.lax.map` or `jax.lax.scan` (after `jax.vmap`)
the results can differ significantly depending on the GPU used and seemingly random chance.

The GPUs I tested this on are 
- Geforce rtx 3080
- Geforce rtx 3060
- A100
- A100 with MIG
- H100

The results vary significantly between gpus and between runs on the same gpu. For each of the tested GPUs, there were runs for which the measured mean absolute error and mean relative error were significantly larger than expected (often both > 1e-4, at times > 1e-2).

The networks I tested were MLPs. Three of them shallow (1 input layer, 1 output layer), two of them deeper (3 and 5 layers in total). The initialization scheme is that described in https://arxiv.org/pdf/2006.09661 and https://arxiv.org/pdf/2102.02611

The python script for a single run was
```python
# this is "numerical_test.py"
from typing import Callable
import jax
from jax import numpy as jnp

lin_space = jnp.linspace(-1, 1, 256)
grid = jnp.stack(jnp.meshgrid(lin_space, lin_space), axis=-1)
print(40*"=")
print("Running Tests:")
# initialization scheme based on SIREN paper
key = jax.random.key(123)
hidden_weights_key, hidden_bias_key, final_weights_key, final_bias_key = jax.random.split(key, 4)
hw_maxval = 1. / 2  # from https://github.com/vsitzmann/siren/blob/4df34baee3f0f9c8f351630992c1fe1f69114b5f/modules.py#L630

hidden_weights = jax.random.uniform(
    key=hidden_weights_key,
    shape=(256, 2),
    minval=-hw_maxval,
    maxval=hw_maxval
)
hidden_bias = jax.random.uniform(
    key=hidden_bias_key,
    shape=(256,),
    minval=-1,
    maxval=1
)
bias_factor = jnp.pi / jnp.sqrt(jnp.sum(jnp.square(hidden_weights), axis=1))  # from https://arxiv.org/pdf/2102.02611.pdf page 6 third paragaph
hidden_bias = bias_factor * hidden_bias

fw_maxval = 1./jnp.sqrt(256)
final_weights = jax.random.uniform(
    key=final_weights_key,
    shape=(1, 256),
    minval=-fw_maxval,
    maxval=fw_maxval
)
final_bias = jax.random.uniform(
    key=final_bias_key,
    shape=(1,),
    minval=-1,
    maxval=1,
)

def shallow_siren(input_coordinate):
    x = input_coordinate
    x = hidden_weights @ x + hidden_bias
    x = jnp.sin(30. * x)

    x = final_weights @ x + final_bias
    return x

def shallow_tanh(input_coordinate):
    x = input_coordinate
    x = hidden_weights @ x + hidden_bias
    x = jnp.tanh(30. * x)

    x = final_weights @ x + final_bias
    return x


def shallow_relu(input_coordinate):
    x = input_coordinate
    x = hidden_weights @ x + hidden_bias
    x = jnp.maximum(30. * x, 0.)

    x = final_weights @ x + final_bias
    return x

def evaluate_on_grid_batch_wise(func:Callable, grid:jax.Array, batch_size:int)->jax.Array:
    """
    evaluate func on grid in a batch-wise fassion

    :param func: a function taking (c,) shaped arrays and returning (c'...) shaped arrays
    :param grid: a (spatial_dims..., c) shaped array of inputs for func
    :param batch_size: the size of each batch func should be evaluated on in parallel
    :return: a (spatial_dims..., c'...) shaped array of outcomes

    func will be vmapped over each batch
    this vmapped func is applied using jax.lax.map
    """
    grid_shape = grid.shape
    batches = jnp.reshape(grid, (-1, batch_size, grid_shape[-1]))
    def evaluate_on_batch(batch):
        return jax.vmap(func)(batch)

    results = jax.lax.map(evaluate_on_batch, batches)
    target_shape = grid_shape[:-1] + results.shape[2:]
    return jnp.reshape(results, target_shape)


def evaluate_on_grid_vmapped(func:Callable, grid:jax.Array)->jax.Array:
    """
    evaluate func on grid in a batch wise fassion

    :param func: a function taking (c,) shaped arrays and returning (c'...) shaped arrays
    :param grid: a (spatial_dims..., c) shaped array of inputs for func
    :return: a (spatial_dims..., c'...) shaped array of outcomes

    func will be vmapped over a reshaped grid
    """
    grid_shape = grid.shape
    flattened_grid = jnp.reshape(grid, (-1,  grid_shape[-1]))

    results = jax.vmap(func)(flattened_grid)
    target_shape = grid_shape[:-1] + results.shape[1:]
    return jnp.reshape(results, target_shape)

def evaluate_on_grid_batch_wise_v2(func:Callable, grid:jax.Array, batch_size:int)->jax.Array:
    """Alternative implementation using scan instead of lax.map"""
    grid_shape = grid.shape
    flattened_grid = jnp.reshape(grid, (-1, grid_shape[-1]))
    slice_sizes = (batch_size, grid.shape[-1])
    def process_batch(carry, batch_start):
        batch_start = [batch_start, jnp.zeros((), jnp.int32)]
        #batch = flattened_grid[batch_start:batch_end]
        batch = jax.lax.dynamic_slice(flattened_grid, batch_start, slice_sizes)
        return carry, jax.vmap(func)(batch)
    
    num_batches = flattened_grid.shape[0] // batch_size
    batch_starts = jnp.arange(num_batches, dtype=jnp.int32) * batch_size
    
    _, results = jax.lax.scan(process_batch, None, batch_starts)
    
    # Flatten and reshape
    flat_results = jnp.reshape(results, (-1, results.shape[-1]))
    target_shape = grid_shape[:-1] + (results.shape[-1],)
    return jnp.reshape(flat_results, target_shape)


print("On GPU:")

def get_difference(network):
    batched_v2 = evaluate_on_grid_batch_wise_v2(
        network,
        grid=grid,
        batch_size=256*8,
    )
    batched_v1 = evaluate_on_grid_batch_wise(
        network,
        grid=grid,
        batch_size=256*8
    )
    non_batched = evaluate_on_grid_vmapped(
        network,
        grid=grid
    )
    abs_difference_v1v2 = jnp.abs(batched_v1 - batched_v2)
    mae_v1v2 = jnp.mean(abs_difference_v1v2)
    abs_difference_v1 = jnp.abs(batched_v1 - non_batched)
    mae_v1 = jnp.mean(abs_difference_v1)
    abs_difference_v2 = jnp.abs(batched_v2 - non_batched)
    mae_v2 = jnp.mean(abs_difference_v2)

    # also get relative errors:
    abs_sizes = jnp.maximum((jnp.abs(batched_v1) + jnp.abs(non_batched))/2, 1e-8)
    mre_v1v2 = jnp.mean(abs_difference_v1v2/abs_sizes)
    mre_v1 = jnp.mean(abs_difference_v1 / abs_sizes)
    mre_v2 = jnp.mean(abs_difference_v2 / abs_sizes)
    print(f"{mae_v1v2=}, {mae_v1=}, {mae_v2=}, {mre_v1v2=}, {mre_v1=}, {mre_v2=}")

print("MAEs and relative errors with siren:")
get_difference(shallow_siren)
print("\nMAEs and relative errors with tanh:")
get_difference(shallow_tanh)
print("\nMAEs and relative errors with relu:")
get_difference(shallow_relu)

print("\n\nNow on CPU:")
with jax.default_device(jax.devices('cpu')[0]):
    print("CPU Results:")
    print("MAEs and relative errors with siren:")
    get_difference(shallow_siren)
    print("\nMAEs and relative errors with tanh:")
    get_difference(shallow_tanh)
    print("\nMAEs and relative errors with relu:")
    get_difference(shallow_relu)

print("\n\n")

print("Tests with deeper SIREN networks")
def make_siren(num_layers, key):
    keys = list(jax.random.split(key, 2*num_layers))
    weights = []
    biases = []
    w0 = 30.

    first_layer_maxval = 1. / 2  # from https://github.com/vsitzmann/siren/blob/4df34baee3f0f9c8f351630992c1fe1f69114b5f/modules.py#L630
    weights.append(jax.random.uniform(
        key=keys.pop(),
        shape=(256, 2),
        minval=-first_layer_maxval,
        maxval=first_layer_maxval
    ))
    bias = jax.random.uniform(
        key=keys.pop(),
        shape=(256,),
        minval=-1,
        maxval=1
    )
    bias_factor = jnp.pi / jnp.sqrt(jnp.sum(jnp.square(weights[-1]), axis=1))  # from https://arxiv.org/pdf/2102.02611.pdf page 6 third paragaph
    bias = bias_factor * bias
    biases.append(bias)

    for _ in range(num_layers-2):
        weights_maxval = jnp.sqrt(6. / 256) / w0 # from https://arxiv.org/pdf/2006.09661.pdf subsection.3.2 and appendix 1.5 and https://github.com/vsitzmann/siren/blob/4df34baee3f0f9c8f351630992c1fe1f69114b5f/modules.py#L627
        weights.append(jax.random.uniform(
            key=keys.pop(),
            shape=(256, 256),
            minval=-weights_maxval,
            maxval=weights_maxval
        ))
        bias = jax.random.uniform(
            key=keys.pop(),
            shape=(256,),
            minval=-1,
            maxval=1
        )
        bias_factor = jnp.pi / jnp.sqrt(jnp.sum(jnp.square(weights[-1]), axis=1))  # from https://arxiv.org/pdf/2102.02611.pdf page 6 third paragaph
        bias = bias_factor * bias
        biases.append(bias)

    fw_maxval = 1./jnp.sqrt(256)
    weights.append(jax.random.uniform(
        key=keys.pop(),
        shape=(1, 256),
        minval=-fw_maxval,
        maxval=fw_maxval
    ))
    biases.append(jax.random.uniform(
        key=keys.pop(),
        shape=(1,),
        minval=-1,
        maxval=1,
    ))

    def siren(x):
        for weight, bias in zip(weights[:-1], biases[:-1]):
            x = weight @ x + bias
            x = jnp.sin(w0 * x)
        return weights[-1] @ x + biases[-1]
    return siren

print("With three layers: ")
get_difference(make_siren(3, jax.random.key(123)))

print("With five layers:")
get_difference(make_siren(5, jax.random.key(123)))

```
I used the following script to get information on the environment in which the scripts were running:
```python
# this is "get_environment_info.py"
import jax
jax.print_environment_info()
```

And I used the following bash script to run 10 tests on a gpu:
```bash
#!/bin/bash
python get_environment_info.py
python numerical_test.py
python numerical_test.py
python numerical_test.py
python numerical_test.py
python numerical_test.py
python numerical_test.py
python numerical_test.py
python numerical_test.py
python numerical_test.py
python numerical_test.py
```
The results can be found in the following .txt files:
[test_on_3060.txt](https://github.com/user-attachments/files/22410685/test_on_3060.txt)
[test_on_3080.txt](https://github.com/user-attachments/files/22410682/test_on_3080.txt)
[R-numerics_test_mig.sh.14778309.txt](https://github.com/user-attachments/files/22410683/R-numerics_test_mig.sh.14778309.txt)
[R-numerics_test_h100.sh.14778420.txt](https://github.com/user-attachments/files/22410681/R-numerics_test_h100.sh.14778420.txt)
[R-numerics_test_a100.sh.14778307.txt](https://github.com/user-attachments/files/22410684/R-numerics_test_a100.sh.14778307.txt)

**Comment by unalmis** [comment]
> on batches of inputs using jax.lax.map or jax.lax.scan (after jax.vmap)

I am not sure what you mean by `after jax.vmap`, but I have also observed that computing a function in small chunk sizes with `jax.lax.map` gives the verifiably correct result while computing it all at once with `jax.vmap` gives wrong and random results. I had made an issue for this, but giving a MWE would be significant effort for my use case, so I was holding out until #27591 was resolved as that may end up fixing my issue. In my use case I could reproduce on CPU tooo but I think it was only when the memory limits were being approached.

**Comment by rukundob451** [comment]
I’ve also noticed discrepancies between jax.vmap and jax.lax.map/scan outputs on GPU, which seem nondeterministic across runs. Could this be tied to nondeterministic GPU kernels or precision trade-offs in certain ops?

**Comment by unalmis** [comment]
I do not think that is the issue because as I increased the batch size in my application I lost all numerical resolution (the output values grew toward +-infinity). If this was simply due to using a less accurate kernel, I'd assume the errors would be bounded since there's probably some sane choice for the minimum precision of an operation.

**Comment by SimonKoop** [comment]
> I am not sure what you mean by after `jax.vmap`

I just meant that I was using a manually vmapped function inside `jax.lax.map`. The code I encountered this problem with is rather old and from before a `batch_size` parameter was introduced in `jax.lax.map` and, tbh, I forgot to change that in the example I posted. Sorry for the vague description.

**Comment by SimonKoop** [comment]
@jakevdp sorry to bother you, but when I wrote this bug report, I messed up and it didn't get the  `bug` label, and now I don't seem to be able to add it myself. Could you maybe add the `bug` label?
